kimchi/src/circuits/polynomials/foreign_field_add/gadget.rs

```diff
@@ -0,0 +1,440 @@
+//! This module obtains the gates of a foreign field addition circuit.
+
+use std::collections::HashMap;
+
+use ark_ff::{FftField, PrimeField, Zero};
+use ark_poly::{
+    univariate::DensePolynomial, EvaluationDomain, Evaluations, Radix2EvaluationDomain as D,
+};
+use rand::{prelude::StdRng, SeedableRng};
+use std::array;
+
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::{Argument, ArgumentType},
+        constraints::ConstraintSystem,
+        expr::{self, l0_1, Environment, LookupEnvironment, E},
+        gate::{CircuitGate, CircuitGateError, CircuitGateResult, Connect, GateType},
+        lookup::{
+            self,
+            lookups::{LookupInfo, LookupsUsed},
+            tables::{GateLookupTable, LookupTable},
+        },
+        polynomial::COLUMNS,
+        wires::Wire,
+    },
+    curve::KimchiCurve,
+};
+
+use super::circuitgates::{ForeignFieldAdd, ForeignFieldFin};
+
+/// Number of gates used by the foreign field addition gadget
+pub const GATE_COUNT: usize = 2;
+
+impl<F: PrimeField> CircuitGate<F> {
+    /// Outputs next row
+    fn append_multi_range_check_rows(next_row: &mut usize, gates: &mut Vec<CircuitGate<F>>) {
+        let (subsequent_row, mut range_check_circuit_gates) =
+            CircuitGate::create_multi_range_check(*next_row);
+        gates.append(&mut range_check_circuit_gates);
+        *next_row = subsequent_row;
+    }
+
+    /// Create foreign field addition gate
+    ///     Inputs
+    ///         starting row
+    ///         number of addition gates
+    ///     Outputs tuple (next_row, circuit_gates) where
+    ///       next_row      - next row after this gate
+    ///       circuit_gates - vector of circuit gates comprising this gate
+    ///
+    /// Note that te final structure of the circuit is as follows:
+    /// circuit_gates = [
+    ///        [0..3]      -> 1 RangeCheck for left_input
+    ///      {
+    ///        [8i+4..8i+7]   -> 1 RangeCheck for right_input
+    ///        [8i+8..8i+11]  -> 1 RangeCheck for result
+    ///      } * num times
+    ///      [8n+4..8n+7]     -> 1 RangeCheck for bound
+    ///      {
+    ///        [8n+i+8] ->    -> 1 ForeignFieldAdd row
+    ///      } * num times
+    ///      [9n+8]           -> 1 ForeignFieldFin row
+    /// ]
+    ///
+    pub fn create_foreign_field_add(start_row: usize, num: usize) -> (usize, Vec<Self>) {
```

Comment:
I think we should add another parameter called `flags` (i.e. an enum type) that allows us to specify which operands are range-checked.  Depending on this flag the gadget can decide whether to add the expensive range checks for the operands.

```rust
enum ForeignFieldRangeCheck {
    None,
    Left,
    Right,
    Both,
}
```

The gadgets create functions can look at these flags to decide which parameters if any need explicit range checks.

Since operands to both `ffAdd` and `ffMul` will often already be range checked it could be common to call this gate without the extra expensive checks, for example like so

```rust
create_foreign_field_add(row, x, ForeignFieldRangeCheck::None)
```

Something like this..

Reply:
I guess this makes sense but it would complicate this PR quite a bit more beause the indices where foreign field addition rows start would not be so clear. Meaning, this would imply some changes also in the witness generation code etc. This is indeed a good idea, but I think it makes more sense to tackle this once ffmul is merged as well and we understand better in what way will these two gates interact with each other. Otherwise I feel like it will be a large amount of work that we may not eventually use as planned...

Reply:
I will intentionally leave this comment unresolved so we can easily go back to it in the near future.